sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -71,8 +85,215 @@ case class ShuffledHashJoinExec(
     val numOutputRows = longMetric("numOutputRows")
     streamedPlan.execute().zipPartitions(buildPlan.execute()) { (streamIter, buildIter) =>
       val hashed = buildHashedRelation(buildIter)
-      join(streamIter, hashed, numOutputRows)
+      joinType match {
+        case FullOuter => fullOuterJoin(streamIter, hashed, numOutputRows)
+        case _ => join(streamIter, hashed, numOutputRows)
+      }
+    }
+  }
+
+  private def fullOuterJoin(
+      streamIter: Iterator[InternalRow],
+      hashedRelation: HashedRelation,
+      numOutputRows: SQLMetric): Iterator[InternalRow] = {
+    val joinKeys = streamSideKeyGenerator()
+    val joinRow = new JoinedRow
+    val (joinRowWithStream, joinRowWithBuild) = {
+      buildSide match {
+        case BuildLeft => (joinRow.withRight _, joinRow.withLeft _)
+        case BuildRight => (joinRow.withLeft _, joinRow.withRight _)
+      }
+    }
+    val buildNullRow = new GenericInternalRow(buildOutput.length)
+    val streamNullRow = new GenericInternalRow(streamedOutput.length)
+    lazy val streamNullJoinRowWithBuild = {
```

Comment:
It's not about `lazy val`. We should do
```
val streamNullJoinRowWithBuild: InternalRow => JoinedRow = buildSide match {
  case BuildLeft => row => {
    joinRow.withRight(streamNullRow)
    joinRow.withLeft(row)
  }
  case ...
}
```

Which is clearer.

Reply:
@cloud-fan - I think with `lazy val` I am to address [comment from @viirya ](https://github.com/apache/spark/pull/29342#discussion_r467252553), to only set stream side NULL row once, but not per row, because every row would have stream side NULL row so we only need to set it once. If not `lazy val`, but `val`, the `joinRow.withRight(streamNullRow)` would be eagerly evaluated here which is not right, as `joinRow` being reused later.